### PR TITLE
chore: relocate events-example from nile-js at version 0.16.0

### PR DESCRIPTION
**Pull request by scottydawg**
This PR relocates the `events-example` from our nile-js repo [as it existed at v0.16.0](https://github.com/TheNileDev/nile-js/tree/fa3d8d10eff7cbd4d138b186e612cce29a4e5f5c/packages/events-example), but with an updated README that more thoroughly explains the example.

### CODE REVIEW

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
+# Nile Events Example #
+
+This example demonstrates how synchronize your data plane and control plane in
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:3]:
```suggestion
This example demonstrates how to synchronize (i.e., `reconcile`) your data plane and control plane in
```

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+real time with Nile events.
+
+Nile doesn't prescribe any particular deployment solution, but here we'll be
+using [Pulumi](https://app.pulumi.com/) to deploy objects into AWS.
+
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:8]: @scottydawg as a followup to [this comment](https://github.com/TheNileDev/nile-js/pull/465#discussion_r954288874), especially because we know there are users not using Pulumi, can you please add what needs to change to use something other than Pulumi?  It doesn't need to be detailed (there could be another example we develop) but at least a hint of what to do/change.

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+## Prerequisites ##
+
+This example assumes you have:
+
+* [An AWS account](https://aws.amazon.com/free/)
+* [A Pulumi account](https://app.pulumi.com/signup) that's
+  [connected to your AWS account](https://www.pulumi.com/docs/get-started/aws/begin/)
+* [The Pulumi CLI installed]()
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:16]:
```suggestion
* [The Pulumi CLI installed](https://www.pulumi.com/docs/reference/cli/)
```

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+* A Nile developer account
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:17]: Is it worth mentioning explicitly that it needs to be a Nile developer account with a password, not passwordless with Google?

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+## Configure Your Control Plane ##
+
+1. [Login to the Nile Admin Dashboard](https://nad.thenile.dev/) If you don't
+   already have one, create a workspace named "cyberdyne".
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:22]: When I went through this workflow, I was coming off the Quickstart.  In anticipation of other users doing the same, would it be acceptable to make the workspace name consistent with the one used in the [quickstart](https://www.thenile.dev/docs/current/quick-start-ui#create-your-workspace), `clustify`?  (If yes, then I believe it requires changes in 3 places in the README.md only)

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      "greeting": {
+        "type": "string"
+      }
+    }
+  }
+}
+```
+
+3. Create an organization in your workspace named "sac-norad".
+4. Create an SkyNet instance in your organization, with a simple greeting:
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:41]:
```suggestion
4. Create a "SkyNet" instance in your organization, with a value that matches the schema defined earlier:
```

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  "greeting": "Come with me if you want to live."
+}
+```
+
+## Configure Your Data Plane ##
+
+> These instructions summarize how to [get started with Pulumi](https://www.pulumi.com/docs/get-started/aws/begin/)
+> on AWS. See their docs for a more complete setup.
+
+Install the Pulumi CLI and set up a new Pulumi project:
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:54]: Pulumi CLI was already installed in the pre-req section.

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+```bash
+[~/] $ brew install pulumi/tap/pulumi
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:57]:
```suggestion
brew install pulumi/tap/pulumi
```

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+[~/] $ mkdir nile-examples && cd nile-examples
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:58]:
```suggestion
mkdir nile-examples && cd nile-examples
```

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+[~/] $ pulumi new aws-typescript
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:59]:
```suggestion
pulumi new aws-typescript
```

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+To create a stack in an organization, use the format <org-name>/<stack-name> (e.g. `acmecorp/dev`).
+stack name: (dev) 
+Created stack 'dev'
+
+aws:region: The AWS region to deploy into: (us-east-1)
+Saved config
+```
+
+Run `pulumi up` to ensure that you've configured Pulumi correctly. This will
+create a stack named `dev`. We won't be using this stack, but its presence
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:79]:
```suggestion
create a new Pulumi stack named `dev`. We won't be using this stack, but its presence
```

File: js/events/deploy-webapp-with-pulumi/.gitignore [hunk 12]
```diff
@@ -0,0 +1,40 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/.gitignore:1]: Maybe add `nile-examples/`

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+Install the Pulumi CLI and set up a new Pulumi project:
+
+```bash
+[~/] $ brew install pulumi/tap/pulumi
+[~/] $ mkdir nile-examples && cd nile-examples
+[~/] $ pulumi new aws-typescript
+```
+
+and accept the defaults:
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:62]: Note that because I already have `nile-examples`, the defaults aren't populating.  I have to rename to get this to work seamlessly.

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+Saved config
+```
+
+Run `pulumi up` to ensure that you've configured Pulumi correctly. This will
+create a stack named `dev`. We won't be using this stack, but its presence
+verifies that you're ready to proceed.
+
+## Create and Execute a Command ##
+
+In the `deploy-webapp-with-pulumi` directory, run `yarn install && yarn build` to
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:84]:
```suggestion
Back up to the `deploy-webapp-with-pulumi` directory and run `yarn install && yarn build` to
```

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+create the executable command binary.
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:85]: Not sure if I'm doing something wrong but this fails:

```
[10:38:05] ~/git/examples/js/events/deploy-webapp-with-pulumi(chore/reloc-events-example) ✗: yarn install && yarn build
yarn install v1.22.19
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
success Already up-to-date.
✨  Done in 0.13s.
yarn run v1.22.19
$ shx rm -rf dist && tsc -b
src/commands/reconcile/index.ts:6:47 - error TS2307: Cannot find module './lib/pulumi' or its corresponding type declarations.

6 import { pulumiS3, PulumiAwsDeployment } from './lib/pulumi';
                                                ~~~~~~~~~~~~~~


Found 1 error.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Review by ybyzek — CHANGES_REQUESTED** [review]
@scottydawg Thanks!  Here is first set of feedback up to a blocker (cannot proceed).

Also, please update the parent README.md to link to this new example

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 15 above)
Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:85]: 😞  The .gitignore in the root of the examples repo ignores all `lib/` directories, so they didn't get included in the PR. I'll figure out how to fix that.

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 2 above)
Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:8]: That's part of what the later revision attempted to do by having a more general `Deployment` interface, and showing the `PulumiAwsDeployment` as a specialization of it. Rather than `loadPulumiStacks()`, `createStack()` and `destroyStack()`, which are all concepts specific to Pulumi, we moved to ` identifyRemoteObjects(), `createObject()` and `destroyObject()` to show that they could be more abstract. Having that interface also gives a little bit of a talking point to be able to say, "You could create a different `Deployment` implementation for Kubernetes, Terraform, etc." It might be worth bringing that one back in the next revision to make this sort of explanation easier.

I'll try to summarize that concisely in the README for now.

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 5 above)
Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:22]: Not a Terminator fan? :)

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 9 above)
Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:58]: Ah, yes. This one touches on a potentially confusing issue. I explicitly included the home directory notation here to indicate that this `nile-examples` dir is independent of this example. I configured my pulumi projects in their own directory, separate from any of my library code, not knowing what the conventional usage is. But maybe we should model this like our `iteru` repo, and have a `/pulumi` dir as a subdir of `/deploy-webapp-with-pulumi`?

Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:58]: (If we go this ☝️ route, we'll probably want to change the subsequent examples to _not_ accept the defaults. Having a project in pulumi named "pulumi" is going to get confusing.)

File: js/events/deploy-webapp-with-pulumi/.gitignore (hunk 12 above)
Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/.gitignore:1]: Based on [this thread](https://github.com/TheNileDev/examples/pull/13#discussion_r956144303), I modeled our structure here on what we do in iteru. So this became `/pulumi`.

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+```bash
+mkdir pulumi && cd pulumi
+pulumi new aws-typescript
+```
+
+and set the project name to "nile-examples":
+
+```
+project name: (pulumi) nile-examples
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:71]: 1. `nile-examples` is no longer valid
2. Strange but the project name defaults to `pulum` (notice the missing `i`) even in the folder `pulumi`.  Maybe call it `pulumi-clustify`?  (Is that reasonable what a user would call their pulumi project to match their workspace?). Or something else...

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+Run `pulumi up` to ensure that you've configured Pulumi correctly. This will
+create a new Pulumi stack named `dev`. We won't be using this stack, but its
+presence verifies that you're ready to proceed.
+
+## Create and Execute a Command ##
+
+Back up in the `deploy-webapp-with-pulumi` directory, run `yarn install && yarn build` 
+to create the executable command binary.
+
+To run the binary, you'll need to pass arguments for your Nile workspace,
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:93]:
```suggestion
To run the `reconcile` binary, you'll need to pass arguments for your Nile workspace,
```

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+>
+> the organization id is `org_02qfJTCBve6bw0XlxC92CG`.
+
+Execute the binary with:
+
+```bash
+./bin/dev reconcile --basePath https://prod.thenile.dev \
+--workspace clustify \
+--entity SkyNet \
+--organization YOUR_ORG_ID \
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:110]: Just noting that I think in my earlier run this was `org`, not `organization`

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+--email YOUR_EMAIL \
+--password YOUR_PASSWORD
+```
+
+This will synchronize your Pulumi stack and instantiate the SkyNet instance
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:115]: It's not clear what `instantiate the SkyNet instance` means.  Can you please elaborate?

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+already defined in your control plane. Check your Pulumi project dashboard to
+confirm.
+
+The command will log out your instance properties, including the `websiteUrl` 
+of the object created by the [`pulumis3` program](./src/pulumi/pulumiS3.ts):
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:120]:
```suggestion
of the object created by the [`pulumis3` program](./src/commands/reconcile/lib/pulumi/pulumiS3.ts):
```

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+```bash
+Outputs:
+
+    bucketPolicy: {
+        // ...redacted...
+    }
+    object      : {
+        // ...redacted...
+    }
+    websiteUrl  : "s3-website-bucket-5c7d7bc.s3-website.us-east-2.amazonaws.com"
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:131]: Should this be redacted?

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+
+Resources:
+    + 4 created
+
+Duration: 5s
+```
+
+Pull up that `websiteUrl` in-browser and you'll find your provided `greeting`
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:139]: It would help someone like me to connect the dots if you added an additional description of what this represents.  Now that a user did all the steps, recap what they've done.  i.e., something like "The `reconcile` binary got the metadata about the event of the new SkyNet instance from Nile and created the appropriate "data plane" by calling Pulumi to create the S3 resources..." (or something like that...

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 21 above)
Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:131]: No, it's a dead bucket and doesn't give away any creds. I left it in place as a sample of what they might expect to see.

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+as well as all of your instance details.
+
+## Add/Remove Instances ##
+
+In your [Nile dashboard](https://nad.thenile.dev/dev/organizations), add one or
+more SkyNet instances to your organization. This will trigger events that the
+command receives, and will synchronize accordingly. Deleting an instance in your
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:146]: IIUC delete is not supported in NAD yet.  If so, maybe reword this to clarify it should be done using the API (if supported).

Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:146]: You can delete in NAD. Select an instance and you'll see a "delete" button.

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 24]
```diff
@@ -0,0 +1,147 @@
+# Nile Events Example #
+
+This example demonstrates how to synchronize (i.e., `reconcile`) your data
+plane and control plane in real time with Nile events.
+
+Nile doesn't prescribe any particular deployment solution, but here we'll be
+using [Pulumi](https://app.pulumi.com/) to deploy objects into AWS. 
+
+> If you're using another tool like Kubernetes or Terraform, you could replace
+> the `PulumiAwsDeployment` class with your own deployment implementation.
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:10]: It would be helpful to tell the user

1. The `reconcile` code is the example here, and that this is what the user needs to provide in their own deployment
2. Where this file is, specifically in `src/commands/reconcile/`

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 23 above)
Comment (ybyzek) [reply js/events/deploy-webapp-with-pulumi/README.md:146]: @scottydawg thank you for the correction, I see it now.

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 24 above)
Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:10]: I'm not sure I follow... If they downloaded the example code, I'm assuming they know that the code is in here. I'll specify the path. (Note that it changes in the next version of this PR.)

File: js/events/deploy-webapp-with-pulumi/README.md [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+
+Duration: 5s
+```
+
+Pull up that `websiteUrl` in-browser and you'll find your provided `greeting`
+as well as all of your instance details.
+
+## Add/Remove Instances ##
+
+In your [Nile dashboard](https://nad.thenile.dev/dev/organizations), add one or
```
Comment (ybyzek) [inline js/events/deploy-webapp-with-pulumi/README.md:144]:
```suggestion
In the [Nile Admin Dashboard](https://nad.thenile.dev/), add one or
```

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 21 above)
Comment (ybyzek) [reply js/events/deploy-webapp-with-pulumi/README.md:131]: @scottydawg makes sense.

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 18 above)
Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:111]: That was an improvement made after 0.16.0. This version requires `organization`.

File: js/events/deploy-webapp-with-pulumi/README.md (hunk 16 above)
Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md:72]: I'm not sure what the pulumi conventions are. I was taking a nod from our iteru repo. Maybe @norwood or @mpokryva knows.

Comment (scottydawg) [reply js/events/deploy-webapp-with-pulumi/README.md]: changed to `pulumi-clustify` in latest.

**Comment by ybyzek** [comment]
@scottydawg thanks for addressing the feedback.  

There are two outstanding issues:

1. I don't think we closed out the discussion about the naming convention for the path where this example sits, so I'd appreciate if @gwenshap were able to weigh in on that first before merging.
2. Please update the [parent README.md](https://github.com/TheNileDev/examples/blob/main/README.md) to link to this new example

**Comment by scottydawg** [comment]
We'll need to settle that path structure before we can update the main README.
cc @gwenshap

**Comment by ybyzek** [comment]
I have renamed the folder to `data-plane/pulumi`, updated the top level README.md, and validated with:


```
 ./bin//dev reconcile --basePath https://prod.thenile.dev \
--workspace foo \
--entity SkyNet \
--organization $MY_ORG_ID \
--email $NILE_DEVELOPER_EMAIL \
--password $NILE_DEVELOPER_PASSWORD
```

**Review by ybyzek — APPROVED** [review]
